ENIGMAsystem/SHELL/Platforms/Cocoa/CocoaMain.cpp

```diff
@@ -21,36 +21,73 @@
 
 #include <unistd.h>
 #include <stdio.h>
+#include <mach-o/dyld.h>
+#include <sys/stat.h>
+#include <limits.h>
 #include "CocoaMain.h"
 #include "ObjectiveC.h"
 #include <cstdlib>
+#include <string>
 
 #include "../General/PFwindow.h"
 #include "../General/PFfilemanip.h"
 
 #include "Platforms/General/PFmain.h"
 #include "Universal_System/roomsystem.h"
 
+using std::string;
+
 namespace enigma {
   void SetResizeFptr();
+  
+  void set_working_directory() {
+    // Set the working_directory
+    char buffer[PATH_MAX + 1];
+    if (getcwd(buffer, PATH_MAX + 1) != NULL)
+      enigma_user::working_directory = buffer;
+
+    // Set the program_directory
+    buffer[0] = 0;
+    char real_executable[PATH_MAX + 1];
+    char *bundle_id;
+
+    uint32_t bufsize = sizeof(buffer);
+    if (_NSGetExecutablePath(buffer, &bufsize) == 0) {
+      bundle_id = dirname(buffer);
+      strcpy(real_executable, bundle_id);
+      strcat(real_executable, "/");
+
+      enigma_user::program_directory = real_executable;
+    }
+
+    // Set the temp_directory
+    char const *env = getenv("TMPDIR");
+
+    if (env == 0)
+      env = "/tmp/";
+
+    enigma_user::temp_directory = env; 
+  }
 }
 
 namespace enigma_user {
-  std::string working_directory = "";
+  double set_working_directory(string dname) {
+    if (!dname.empty()) {
+      while (*dname.rbegin() == '/') {
+        dname.erase(dname.size() - 1);
+      }
+    }
+
+    struct stat sb;
+    if (stat((char *)dname.c_str(), &sb) == 0 && S_ISDIR(sb.st_mode))
+      return (chdir((char *)dname.c_str()) == 0);
+
+    return 0;
+  }
```

Comment:
Please don't duplicate this block from POSIX/main.cpp which it is 100% identical to.

Reply:
You extracted this to `Platforms/General/POSIX/POSIXworkdir.cpp` which is acceptable. Thank you.